File: 14-TypeScript/Grey Multi-Tenant/packages/grey-julia/src/GreyError.jl
"""
Normalized error type for the Grey SDK.
"""
module GreyError

export Error, 
       unauthorized, forbidden, not_found, validation, 
       network, timeout, server, unknown,
       from_response, from_exception

using ..ErrorCodes

"""
Represents a normalized error in the Grey SDK.

# Fields
- `code::Symbol`: Error code
- `message::String`: Error message
- `details::Union{Dict, Nothing}`: Additional error details
"""
struct Error
    code::Symbol
    message::String
    details::Union{Dict{String, Any}, Nothing}
    
    function Error(code::Symbol, message::String, details::Union{Dict, Nothing}=nothing)
        new(ErrorCodes.normalize(code), message, details)
    end
end

"""Create an unauthorized error."""
function unauthorized(message::String="Authentication required")::Error
    Error(ErrorCodes.UNAUTHORIZED, message)
end

"""Create a forbidden error."""
function forbidden(message::String="Permission denied")::Error
    Error(ErrorCodes.FORBIDDEN, message)
end

"""Create a not found error."""
function not_found(message::String="Resource not found")::Error
    Error(ErrorCodes.NOT_FOUND, message)
end

"""Create a validation error."""
function validation(message::String, details::Union{Dict, Nothing}=nothing)::Error
    Error(ErrorCodes.VALIDATION_ERROR, message, details)
end

"""Create a network error."""
function network(message::String="Network error occurred")::Error
    Error(ErrorCodes.NETWORK_ERROR, message)
end

"""Create a timeout error."""
function timeout(message::String="Request timed out")::Error
    Error(ErrorCodes.TIMEOUT, message)
end

"""Create a server error."""
function server(message::String="Server error occurred")::Error
    Error(ErrorCodes.SERVER_ERROR, message)
end

"""Create an unknown error."""
function unknown(message::String="An unknown error occurred")::Error
    Error(ErrorCodes.UNKNOWN, message)
end

"""
    from_response(status::Int, body::Dict) -> Error

Create an error from an HTTP response.
"""
function from_response(status::Int, body::Union{Dict, Nothing}=nothing)::Error
    code = ErrorCodes.from_http_status(status)
    message = if body !== nothing
        get(body, "message", get(body, "error", "HTTP error $status"))
    else
        "HTTP error $status"
    end
    details = body !== nothing ? get(body, "details", nothing) : nothing
    
    Error(code, message, details)
end

"""
    from_exception(ex::Exception) -> Error

Create an error from an exception.
"""
function from_exception(ex::Exception)::Error
    msg = string(ex)
    
    if occursin(r"timeout"i, msg)
        timeout(msg)
    elseif occursin(r"connect|connection|network"i, msg)
        network(msg)
    else
        unknown(msg)
    end
end

end # module

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-matlab/+grey/+error/GreyError.m
classdef GreyError
    %GREYERROR Normalized error type for Grey SDK
    %   Represents an error with code, message, and optional details.
    
    properties (SetAccess = private)
        code char      % Error code
        message char   % Error message
        details        % Additional error details (struct or empty)
    end
    
    methods
        function obj = GreyError(code, message, details)
            %GREYERROR Construct an error
            %   err = GreyError(code, message)
            %   err = GreyError(code, message, details)
            
            obj.code = grey.error.ErrorCodes.normalize(code);
            obj.message = message;
            
            if nargin < 3
                obj.details = [];
            else
                obj.details = details;
            end
        end
    end
    
    methods (Static)
        function err = unauthorized(message)
            %UNAUTHORIZED Create unauthorized error
            if nargin < 1
                message = 'Authentication required';
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.UNAUTHORIZED, message);
        end
        
        function err = forbidden(message)
            %FORBIDDEN Create forbidden error
            if nargin < 1
                message = 'Permission denied';
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.FORBIDDEN, message);
        end
        
        function err = notFound(message)
            %NOTFOUND Create not found error
            if nargin < 1
                message = 'Resource not found';
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.NOT_FOUND, message);
        end
        
        function err = validation(message, details)
            %VALIDATION Create validation error
            if nargin < 2
                details = [];
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.VALIDATION_ERROR, message, details);
        end
        
        function err = network(message)
            %NETWORK Create network error
            if nargin < 1
                message = 'Network error occurred';
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.NETWORK_ERROR, message);
        end
        
        function err = timeout(message)
            %TIMEOUT Create timeout error
            if nargin < 1
                message = 'Request timed out';
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.TIMEOUT, message);
        end
        
        function err = server(message)
            %SERVER Create server error
            if nargin < 1
                message = 'Server error occurred';
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.SERVER_ERROR, message);
        end
        
        function err = unknown(message)
            %UNKNOWN Create unknown error
            if nargin < 1
                message = 'An unknown error occurred';
            end
            err = grey.error.GreyError(grey.error.ErrorCodes.UNKNOWN, message);
        end
        
        function err = fromResponse(status, body)
            %FROMRESPONSE Create error from HTTP response
            %   err = GreyError.fromResponse(status, body)
            
            code = grey.error.ErrorCodes.fromHttpStatus(status);
            
            if isstruct(body) && isfield(body, 'message')
                message = body.message;
            elseif isstruct(body) && isfield(body, 'error')
                message = body.error;
            else
                message = sprintf('HTTP error %d', status);
            end
            
            if isstruct(body) && isfield(body, 'details')
                details = body.details;
            else
                details = [];
            end
            
            err = grey.error.GreyError(code, message, details);
        end
        
        function err = fromException(ex)
            %FROMEXCEPTION Create error from MATLAB exception
            %   err = GreyError.fromException(ex)
            
            msg = ex.message;
            
            if contains(lower(msg), 'timeout')
                err = grey.error.GreyError.timeout(msg);
            elseif contains(lower(msg), {'connect', 'connection', 'network'})
                err = grey.error.GreyError.network(msg);
            else
                err = grey.error.GreyError.unknown(msg);
            end
        end
    end
end
